File: lib/env-validation.ts
/**
 * Environment Variable Validation
 * 
 * Validates that all required environment variables are set before the app starts.
 * This helps catch configuration errors early in production.
 */

interface RequiredEnvVars {
  [key: string]: string | undefined;
}

interface ValidationResult {
  isValid: boolean;
  missingVars: string[];
  warnings: string[];
}

/**
 * Validate required environment variables
 */
export function validateEnvironment(): ValidationResult {
  const missingVars: string[] = [];
  const warnings: string[] = [];
  
  // Critical environment variables
  const requiredVars: RequiredEnvVars = {
    'NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY': process.env.NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY,
    'CLERK_SECRET_KEY': process.env.CLERK_SECRET_KEY,
  };
  
  // Highly recommended variables (warnings only)
  const recommendedVars: RequiredEnvVars = {
    'NEXT_PUBLIC_PYTHON_API_URL': process.env.NEXT_PUBLIC_PYTHON_API_URL,
    'TINYBIRD_TOKEN': process.env.TINYBIRD_TOKEN,
    'OPENAI_API_KEY': process.env.OPENAI_API_KEY,
  };
  
  // Check required variables
  for (const [key, value] of Object.entries(requiredVars)) {
    if (!value || value === '') {
      missingVars.push(key);
    }
  }
  
  // Check recommended variables (only in production)
  if (process.env.NODE_ENV === 'production') {
    for (const [key, value] of Object.entries(recommendedVars)) {
      if (!value || value === '') {
        warnings.push(`Recommended environment variable missing: ${key}`);
      }
    }
  }
  
  return {
    isValid: missingVars.length === 0,
    missingVars,
    warnings,
  };
}

/**
 * Validate and throw error if critical variables are missing
 */
export function validateOrThrow(): void {
  const result = validateEnvironment();
  
  if (!result.isValid) {
    const errorMessage = `
❌ Missing required environment variables:
${result.missingVars.map(v => `  - ${v}`).join('\n')}

Please check your .env file and ensure all required variables are set.
See .env.example for a template.
    `.trim();
    
    throw new Error(errorMessage);
  }
  
  // Log warnings
  if (result.warnings.length > 0) {
    console.warn('⚠️  Environment warnings:');
    result.warnings.forEach(warning => console.warn(`  - ${warning}`));
  }
}

/**
 * Check if we're running in production mode
 */
export function isProduction(): boolean {
  return process.env.NODE_ENV === 'production';
}

/**
 * Check if we're running in development mode
 */
export function isDevelopment(): boolean {
  return process.env.NODE_ENV === 'development' || !process.env.NODE_ENV;
}

File: apps/backend/config/env_validation.py
"""
Environment Variable Validation

Validates that all required environment variables are set before the backend starts.
This helps catch configuration errors early in production.
"""

import os
import sys
import logging
from typing import List, Dict, Tuple

logger = logging.getLogger(__name__)

def validate_environment() -> Tuple[bool, List[str], List[str]]:
    """
    Validate required environment variables
    
    Returns:
        Tuple of (is_valid, missing_vars, warnings)
    """
    missing_vars: List[str] = []
    warnings: List[str] = []
    
    # Critical environment variables
    required_vars = [
        'NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY',
        'CLERK_SECRET_KEY',
        'CLERK_JWKS_URL',
        'CLERK_WEBHOOK_SIGNING_SECRET',
    ]

    is_production_environment = (
        os.getenv('NODE_ENV') == 'production'
        or os.getenv('DEBUG', 'true').lower() == 'false'
    )
    if is_production_environment:
        required_vars.extend([
            'CLERK_ISSUER',
            'CLERK_AUTHORIZED_PARTIES',
            'TURSO_PLATFORM_API_TOKEN',
            'TURSO_ORGANIZATION',
            'TURSO_GROUP',
        ])

    # Highly recommended variables (warnings only)
    recommended_vars = [
        'DATABASE_URL',
        'TINYBIRD_TOKEN',
        'OPENAI_ANSWER_MODEL',
    ]
    
    # Check required variables
    for var in required_vars:
        if not os.getenv(var):
            missing_vars.append(var)
    
    # Check recommended variables (only in production)
    if is_production_environment:
        for var in recommended_vars:
            if not os.getenv(var):
                warnings.append(f"Recommended environment variable missing: {var}")
    
    is_valid = len(missing_vars) == 0
    return is_valid, missing_vars, warnings


def validate_or_exit() -> None:
    """
    Validate and exit if critical variables are missing
    """
    is_valid, missing_vars, warnings = validate_environment()
    
    if not is_valid:
        error_message = f"""
❌ Missing required environment variables:
{chr(10).join(f'  - {v}' for v in missing_vars)}

Please check your .env file and ensure all required variables are set.
See .env.example for a template.
        """.strip()
        
        logger.error(error_message)
        sys.exit(1)
    
    # Log warnings
    if warnings:
        logger.warning("Environment warnings:")
        for warning in warnings:
            logger.warning("  - %s", warning)


def is_production() -> bool:
    """Check if we're running in production mode"""
    return os.getenv('NODE_ENV') == 'production' or os.getenv('DEBUG', 'true').lower() == 'false'


def is_development() -> bool:
    """Check if we're running in development mode"""
    return not is_production()
